Add support for admission controller based on namespace node selectors.

aveshagarwal opened the pull request:
This work is to upstream openshift's project node selectors based admission controller.

Fixes https://github.com/kubernetes/kubernetes/issues/17151

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/24980)

<!-- Reviewable:end -->

aveshagarwal commented:
@kubernetes/rh-cluster-infra

aveshagarwal commented:
Added test cases for namespace cahce and labels.

DirectXMan12 commented on pkg/labels/selector.go:
```diff
@@ -809,3 +809,73 @@ func SelectorFromSet(ls Set) Selector {
+}
+
+// Merge combines given maps
+// Note: It doesn't not check for any conflicts between the maps
```
You should note that the second map `wins` when conflicts do occur.

aveshagarwal replied:
Here it is just to find out if a conflict occurs (true or false) that means if  a same key exists, then their values are same or not. Not sure what do you mean by win here?

aveshagarwal replied:
@DirectXMan12 nv I was thinking about conflicts.

DirectXMan12 commented on pkg/labels/selector.go:
```diff
@@ -809,3 +809,73 @@ func SelectorFromSet(ls Set) Selector {
+}
+
+// Converts selector string to labels map and validates keys and values
+func ConvertSelectortoLabelsMap(selector string) (map[string]string, error) {
```
Nit: should be ConvertSelector**To**LabelsMap

DirectXMan12 commented on pkg/labels/selector.go:
```diff
@@ -809,3 +809,73 @@ func SelectorFromSet(ls Set) Selector {
+	return true
+}
+
+// Converts selector string to labels map and validates keys and values
```
Nit: Doc should be in the form of "ConvertSelectorToLabelsMap frobs the bits..." (start with the method name).

aveshagarwal commented:
@DirectXMan12 thanks. Updated.

aveshagarwal commented:
I noticed that selectors like "a=b, c=d, x=y" are valid as it does not have to be "a=b,c=d,x=y" always. So added white space trimming and now it brings it to parity with the openshift implementation w.r.t label processing. Also added more unit tests.

ncdc commented on pkg/labels/selector.go:
```diff
@@ -809,3 +809,77 @@ func SelectorFromSet(ls Set) Selector {
 	return parse(selector)
 }
+
+// Conflicts takes 2 maps
```
When you view the godoc as rendered html in a web browser, it shows up like this:

"Conflicts takes 2 maps returns true if there a key match between the maps but the value doesn't match returns false in other cases"

It runs together. Please update accordingly.

ncdc replied:
This also only checks the keys in the first map. If you have keys in the second map that don't exist in the first, this function doesn't consider that a conflict.

aveshagarwal replied:
yeah because that is not a conflict, only same keys cause conflict so iterating on one map is fine.

ncdc commented on pkg/labels/selector.go:
```diff
@@ -809,3 +809,77 @@ func SelectorFromSet(ls Set) Selector {
+	return false
+}
+
+// Merge combines given maps
```
Godoc needs formatting update

ncdc commented on pkg/labels/selector.go:
```diff
@@ -809,3 +809,77 @@ func SelectorFromSet(ls Set) Selector {
+func ConvertSelectorToLabelsMap(selector string) (map[string]string, error) {
+	labelsMap := map[string]string{}
+
+	if len(selector) != 0 {
```
```go
if len(selector) == 0 {
  return labelsMap
}
...
```

ncdc commented on plugin/pkg/admission/nodeenv/interfaces.go:
```diff
@@ -0,0 +1,25 @@
+	"k8s.io/kubernetes/pkg/api"
+)
+
+type NamespaceCache interface {
```
NamespaceGetter

ncdc replied:
Godoc

ncdc commented on plugin/pkg/admission/nodeenv/interfaces.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+type NamespaceCache interface {
+	GetNamespace(name string) (*api.Namespace, error)
```
Godoc

ncdc commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,117 @@
+package nodeenv
```
Wait to do this at the end of the reviews, but I want to pick a different name for this package.

smarterclayton replied:
The name for this plugin should be podplacementrestriction or namespacepodplacement or something.

smarterclayton replied:
Please change the package name to be more appropriate.  Env doesn't really make sense in this context.

derekwaynecarr replied:
can we move the package to `plugin/pkg/admission/namespace/nodeselector` ?

derekwaynecarr replied:
or `plugin/pkg/admission/namespace/podnodeselector`

aveshagarwal replied:
going with podnodeselector .

ncdc commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,117 @@
+
+	obj := a.GetObject()
+	pod, ok := obj.(*api.Pod)
+	if !ok {
```
Add a log message that you expected a pod but got a ...

ncdc commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,117 @@
+	}, nil
+}
+
+func (p *podNodeEnvironment) DefaultGetNamespace(name string) (*api.Namespace, error) {
```
do not export

ncdc commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,117 @@
+	return namespace, nil
+}
+
+func GetNodeSelectorMap(namespace *api.Namespace) (map[string]string, error) {
```
do not export

ncdc commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,117 @@
+	return nil
+}
+
+func NewPodNodeEnvironment(client clientset.Interface, nsCache NamespaceCache) (admission.Interface, error) {
```
I think it makes sense to model this after https://github.com/kubernetes/kubernetes/blob/0061479890ddf827bd55570ccbf8bd59352543c3/plugin/pkg/admission/namespace/lifecycle/admission.go#L114, i.e. just take in the client, and create a store/reflector in the body of the method.

aveshagarwal replied:
I remember based on our previous discussion with @deads2k  that we do not want to create its own reflectors here. Not sure when this changed. Previous revision of this PR was doing it similarly only in a different file.  Anyway I have modified it and will send update soon.

ncdc replied:
We'll do a reflector in a follow-up PR. It requires additional plumbing that we don't have right now in Kube.

ncdc commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,117 @@
+	return &podNodeEnvironment{
+		Handler: admission.NewHandler(admission.Create),
+		client:  client,
+		nsCache: nsCache,
```
Create a store above and set it here, like this: https://github.com/kubernetes/kubernetes/blob/0061479890ddf827bd55570ccbf8bd59352543c3/plugin/pkg/admission/namespace/lifecycle/admission.go#L115-L129

ncdc commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,117 @@
+
+const (
+	NamespaceNodeSelector = "kubernetes.io/node-selector"
+	DefaultNodeSelector   = ""
```
This should be something the cluster admin can set when configuring the cluster. It needs to come from `config` on line 15 above

ncdc commented on plugin/pkg/admission/nodeenv/admission_test.go:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	tests := []struct {
+		namespaceNodeSelector           string
```
add a field for defaultNodeSelector

ncdc commented on plugin/pkg/admission/nodeenv/admission_test.go:
```diff
@@ -0,0 +1,198 @@
+			podNodeSelector:       map[string]string{},
+			mergedNodeSelector:    map[string]string{"infra": "false"},
+			admit:                 true,
+			testName:              "TestNamespace node selector and no conflicts",
```
Isn't this case the same as line 48?

ncdc commented on plugin/pkg/admission/nodeenv/admission_test.go:
```diff
@@ -0,0 +1,198 @@
+			podNodeSelector:       map[string]string{},
+			mergedNodeSelector:    map[string]string{"infra": "true"},
+			admit:                 true,
+			testName:              "Default and namespace node selector, no conflicts",
```
Isn't this case the same as line 48 (except instead of infra=false, this is infra=true).

ncdc commented on plugin/pkg/admission/nodeenv/admission_test.go:
```diff
@@ -0,0 +1,198 @@
+		{
+			namespaceNodeSelector: "infra=true",
+			podNodeSelector:       map[string]string{"infra": "false"},
+			mergedNodeSelector:    map[string]string{"infra": "false"},
```
I would omit this since admit=false - it's a bit confusing to see it here (and add a `continue` after line 111 if !test.admit)

aveshagarwal commented:
Updated this PR.

derekwaynecarr commented:
I will review this tomorrow.

bgrant0607 commented:
cc @davidopp

derekwaynecarr commented:
I have yet to do a review on this, may not make top of my queue until tomorrow.

derekwaynecarr commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,146 @@
+	*admission.Handler
+	client              clientset.Interface
+	store               cache.Store
+	defaultNodeSelector string
```
can you add a comment what this value is?

aveshagarwal replied:
yeah sure.

aveshagarwal replied:
Sorry, not sure you meant to add the comment in the code or here, 

Anyway defaultNodeSelector is the global default value (currently default to "") that is assigned to a name space if it does not have any at the time of its creation.

ncdc replied:
The default can't be hardcoded to "" above. It needs to come from the config.

derekwaynecarr replied:
I meant in the code... I also was trying to work out why the default was ""

derekwaynecarr replied:
yeah, so that was not obvious.  maybe change the name to 'clusterDefaultNodeSelector' and add a comment to that effect.

aveshagarwal replied:
ok, if we want to get it from config, do we want to add a command line flag too to allow admins to configure the default?

ncdc commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,146 @@
+	obj := a.GetObject()
+	pod, ok := obj.(*api.Pod)
+	if !ok {
+		glog.Warning("expected pod but got something else")
```
Say what type the something else is

ncdc replied:
Also this is an Errorf, not a Warning

derekwaynecarr commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,146 @@
+
+func NewPodNodeEnvironment(client clientset.Interface, defaultNodeSelector string) (admission.Interface, error) {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
```
can we get a // TODO here to make this a shared cache?

derekwaynecarr commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,146 @@
+}
+
+const (
+	NamespaceNodeSelector = "kubernetes.io/node-selector"
```
I think we need to have a priority list of annotations to pull from to support our downstream usage previously using openshift.io/node-selector.

Can the `podNodeEnvironment` get configured with the slice of annotations to look at?  That way when this is started in OpenShift, we can have a single admission controller that is configured to look at both "kubernetes.io/node-selector" and "openshift.io/node-selector"?

aveshagarwal replied:
Yeah that sounds good to me.

derekwaynecarr commented:
cc @davidopp @bgrant0607 - this is basically upstreaming the feature demonstrated here: 

https://blog.openshift.com/deploying-applications-to-specific-nodes/

davidopp commented:
Is this use case for this "dedicated nodes"?

Taints/tolerations PR is almost finished (#24134) and there will be an admission controller that assigns tolerations based on annotation on namespace. So if this is for the dedicated node use case, we will already have that covered using that mechanism. Not that it's the end of the world to have two approaches doing the same thing, but the taints/toleration approach is more general so we will probably want to encourage people to use that long-term.

I'm OOO through Friday so will have extremely limited bandwidth to discuss this.

aveshagarwal commented:
I have incorporated latest comments by @derekwaynecarr @ncdc, and updated this PR. PTAL.

aveshagarwal commented:
Fixed tests.

sttts commented on pkg/labels/selector.go:
```diff
@@ -806,3 +806,83 @@ func SelectorFromSet(ls Set) Selector {
+		labels := strings.Split(selector, ",")
+		for _, label := range labels {
+			l := strings.Split(label, "=")
+			if len(l) == 2 {
```
same here:
```
if len(l) != 2 {
  return ...
}
```

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,187 @@
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+var NamespaceNodeSelectors = []string{"kubernetes.io/node-selector"}
```
Public variables are bad style.

deads2k replied:
> Public variables are bad style.

This needs to be available as a plug point downstream for us to support our existing annotation.  Kube lacks the config concept like https://github.com/openshift/origin/blob/master/pkg/cmd/server/api/types.go#L1175, so actually using the controller config file is comparatively difficult.

smarterclayton replied:
This needs to be in a policy namespace (scoped).

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+type pluginConfig struct {
+	PodNodeEnvironmentPluginConfig map[string]string
```
is it necessary to make `PodNodeEnvironmentPluginConfig` public? Type is private.

ncdc replied:
For json/yaml decoding

deads2k replied:
> For json/yaml decoding

Unversioned without a scheme?

deads2k replied:
> Unversioned without a scheme?

Ah, you're having to do this because kube still has a single file that's trying to be the config for every admission plugin so you can't have separate apiVersion/kind?

aveshagarwal replied:
separate apiVersion/kind just for configuring default cluster value for a plugin?

deads2k replied:
> separate apiVersion/kind just for configuring default cluster value for a plugin?

Yes, different configs and objects are different.  Add to that the fact that you probably want a field for choosing annotations that has conditional defaulting and it starts looking a lot like an API object.

deads2k replied:
> separate apiVersion/kind just for configuring default cluster value for a plugin?

Unfortunately, since kube works on a single shared controller config file, you have no reasonable way to manage that.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,187 @@
+// The format in a file:
+// podNodeEnvironmentPluginConfig:
+//  clusterDefaultNodeSelector: <node-selectors-labels>
+func readConfig(config io.Reader) string {
```
Feels strange to return a string here. Why not make it `pluginConfig`? That would make the big godoc unnecessary above.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,187 @@
+	if err != nil {
+		return errors.NewInternalError(err)
+	}
+
```
blank line doesn't make sense here.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,187 @@
+	if len(namespace.ObjectMeta.Annotations) > 0 {
+		for _, annotation := range NamespaceNodeSelectors {
+			if ns, ok := namespace.ObjectMeta.Annotations[annotation]; ok {
+
```
blank line unnecessary

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,187 @@
+		store,
+		0,
+	)
+	reflector.Run()
```
How about following https://github.com/kubernetes/kubernetes/blob/4970f0c02d5384763b9833a35c639cea95df1e0f/plugin/pkg/admission/security/podsecuritypolicy/admission.go#L49 having `Run()` outside of the constructor. This looks like leaking go routines, plus without stop channel.

aveshagarwal commented:
Updated based on feedback, rebased and fixed unit tests.
@derekwaynecarr @sttts @ncdc PTAL

aveshagarwal commented:
Fixed failing tests and updated this PR.

aveshagarwal commented:
@derekwaynecarr PTAL.

derekwaynecarr commented:
@aveshagarwal - I am going to revisit this once kube 1.3 is cut since it wont make this milestone.

aveshagarwal commented:
@derekwaynecarr rebased and retested PTAL.

aveshagarwal commented:
@k8s-bot Please keep it open.

aveshagarwal commented:
Rebased this PR, updated it with shared informer cache, implemented whitelist verification and added new units tests. @derekwaynecarr  PTAL.

smarterclayton commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,211 @@
+	"k8s.io/kubernetes/pkg/util/yaml"
+)
+
+var NamespaceNodeSelectors = []string{"kubernetes.io/node-selector"}
```
Going to have to be `alpha.pod` or `pod.alpha`.

aveshagarwal replied:
Based on other annotation keys in pkg/api/helpers.go, shouldn't it be scheduler.alpha.kubernetes.io/node-selector like scheduler.alpha.kubernetes.io/tolerations and scheduler.alpha.kubernetes.io/affinity?

smarterclayton commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,211 @@
+		}
+	}
+	if !found {
+		selector, err = labels.ConvertSelectorToLabelsMap(p.clusterNodeSelectors["clusterDefaultNodeSelector"])
```
This should be a structured type, not a map[string].

aveshagarwal replied:
But this requires to be merged with pod.Spec.NodeSelector which is map[string]string?

derekwaynecarr replied:
look at labels.Set?  i think that would be cleaner to use.

derekwaynecarr commented:
I'll play devils advocate, why alpha?  What would we really do to promote
this from alpha to beta to final form?  Do we really see the final form
being promoted directly to the NamespaceSpec?

On Thursday, September 22, 2016, Avesh Agarwal <notifications@github.com>
wrote:

> *@aveshagarwal* commented on this pull request.
> ------------------------------
>
> In plugin/pkg/admission/nodeenv/admission.go
> <https://github.com/kubernetes/kubernetes/pull/24980>:
>
> > +	"io"
> +	"reflect"
> +
> +	"github.com/golang/glog"
> +
> +	"k8s.io/kubernetes/pkg/admission"
> +	"k8s.io/kubernetes/pkg/api"
> +	"k8s.io/kubernetes/pkg/api/errors"
> +	"k8s.io/kubernetes/pkg/client/cache"
> +	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
> +	"k8s.io/kubernetes/pkg/controller/informers"
> +	"k8s.io/kubernetes/pkg/labels"
> +	"k8s.io/kubernetes/pkg/util/yaml"
> +)
> +
> +var NamespaceNodeSelectors = []string{"kubernetes.io/node-selector"}
>
> Based on other annotation keys in pkg/api/helpers.go, shouldn't it be
> scheduler.alpha.kubernetes.io/node-selector like
> scheduler.alpha.kubernetes.io/tolerations and
> scheduler.alpha.kubernetes.io/affinity?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24980>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF8dbDwkBstNnIom2QIf1L2jCBv-2euDks5qsreqgaJpZM4IS375>
> .
>

aveshagarwal commented:
@derekwaynecarr +1

aveshagarwal commented:
@derekwaynecarr @smarterclayton so alpha or not alpha? I agree with Derek regarding why alpha  as we have not discussed any path forward yet to take it beyond alpha if we end up doing it alpha. Or it is a general practice to set it alpha as we are using annotations? Also any other feedback?

aveshagarwal commented:
rebased.

derekwaynecarr commented:
@aveshagarwal -- let's stick with alpha now.

aveshagarwal commented:
@derekwaynecarr so which one: scheduler.alpha.kubernetes.io/node-selector  or pod.alpha.kubernetes.io/node-selector ?

derekwaynecarr commented on pkg/labels/selector.go:
```diff
@@ -820,3 +820,101 @@ func SelectorFromValidatedSet(ls Set) Selector {
+
+// Conflicts takes 2 maps and returns true if there a key match between
+// the maps but the value doesn't match, and returns false in other cases
+func Conflicts(labels1, labels2 map[string]string) bool {
```
should this be `labels.Set` ?

derekwaynecarr commented on pkg/labels/selector.go:
```diff
@@ -820,3 +820,101 @@ func SelectorFromValidatedSet(ls Set) Selector {
+
+// Merge combines given maps, and does not check for any conflicts
+// between the maps. In case of conflicts, second map (labels2) wins
+func Merge(labels1, labels2 map[string]string) map[string]string {
```
same here, just make this `labels.Set`

derekwaynecarr replied:
the same comment applies throughout.

derekwaynecarr reviewed (changes requested):
few more updates then lgtm

Commit:
Add support for admission controller based on namespace node selectors.

This work is to upstream openshift's project node selectors based admission controller.

Addresses #17151.

Commit:
Added new unit tests.

aveshagarwal commented:
@k8s-bot kubemark e2e test this

aveshagarwal commented:
@derekwaynecarr updated it. PTAL.

aveshagarwal commented:
The following errors dont seem related to a flake.
@k8s-bot kubemark e2e test this

aveshagarwal commented:
@derekwaynecarr @smarterclayton Would appreciate if you could tell me if anything else is needed to get it in?

davidopp commented:
Per our discussion at the SIG meeting today, this doesn't need an LGTM from me, feel free to merge once the appropriate RedHat people are happy with it.

derekwaynecarr commented:
Reviewing this now.

derekwaynecarr commented:
LGTM

aveshagarwal commented:
@k8s-bot test this issue: #IGNORE

aveshagarwal commented:
the pr https://github.com/kubernetes/kubernetes/pull/32555 merged and changed the signature of admission.NewPluginInitializer which is now breaking the build of this admission controller. Will fix it soon.

aveshagarwal commented:
@derekwaynecarr  this pr needs LGTM again.
